Bug Fix in Ladybug Shading Mask II

devngc opened the pull request:
The component is not outputting any geometry from its maskedSkyDome output

mostaphaRoudsari commented:
@devngc are you making new changes to the WindRose component too?

devngc commented:
@mostaphaRoudsari,

I sorry for this. I missed it. I am not sure how windrose files ended up here. Can we close this PR? I can make fresh one instantly. I only meant to send revisions in Shading Mask II component in response to #319 

Devang

mostaphaRoudsari commented on src/Ladybug_Shading Mask_II.py:
```diff
@@ -294,27 +295,34 @@ def main(cenPt, context, radius, merge):
+
+    # sorting the dictionary so that the brep with the largest zValue remains last in the
+    # dictionary
+    sortedDict = sorted(brepDict.items(), key = operator.itemgetter(1))
```
@devngc just use OrderedDict from collections so you don't have to sort the values which are already sorted.

mostaphaRoudsari commented on src/Ladybug_Shading Mask_II.py:
```diff
@@ -294,27 +295,34 @@ def main(cenPt, context, radius, merge):
+    # dictionary
+    sortedDict = sorted(brepDict.items(), key = operator.itemgetter(1))
+    # The brep with smallest zValue for srfCenPt is assumed to be the unmaskedSkyDome
+    unmaskedSkyDome.append(sortedDict[0][0])
```
Why do you make this assumption? Is this the ground object? otherwise why it should be always unmasked?

devngc replied:
@mostaphaRoudsari ,

The component was not outputting anything from it's maskedSkyDome output. When I looked, I found out that all the the breps were going to the unmaskedSkyDome list. Therefore, I came up with this idea to separate those breps.

If the user is using a regular shading geometry, It will not be too close to the cenPt. So the brep with SrfCenPt with highest Z value is going to be the the brep to be supplied to the maskedSkyDome.

Devang

devngc replied:
@mostaphaRoudsari,

I ran a few scenarios. The assumption holds up.

mostaphaRoudsari commented on src/Ladybug_Shading Mask_II.py:
```diff
@@ -294,27 +295,34 @@ def main(cenPt, context, radius, merge):
+    i = 1
+    while i < len(sortedDict):
+        maskedSkyDome.append(sortedDict[i][0])
+        i += 1
```
maskedSkyDome = tuple(sortedDict[key][0] for key in sortedDict.iterkeys())

devngc commented:
@mostaphaRoudsari 

I looked into your suggestions. While using OrderedDict works, it returns a dict. My method of separating breps requires me to access items by index and thefore, a list is easier to work with. 

If you think using sortedlist will somehow slow the performance of this component and it is imperative that I use OrderedDict then please let me know. I will work more on it.

Please find three scenarios with shading surface at different height. The implementation identifies correct maskedSkyDome each time.

1. Best Case Scenario: Shading surface is too away from the testPt
![best case_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20357090/9e8528d2-ac4c-11e6-9bbc-a247ef3f0743.png)

2. Normal Case Scenario: Shading surface is at a good distance from the testPt. This is like a bus stop roof scenario.
![normal case_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20357132/c1626c52-ac4c-11e6-9054-3b0108851ee7.png)

3. 1. Worst Case Scenario: Shading surface is close to the testPt. A user is very unlikely to do this in my view.
![worst case_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20357161/dfd21a5c-ac4c-11e6-81ae-5dc8f85cb7ae.png)

Devang

devngc commented:
Hi @mostaphaRoudsari ,

In the latest commit, I have removed the seam from the skyDome as recommended by Antonello. Also,  parsing the surfaces solely based on points closest to the centroid was supplying all the surfaces to either unmaskedSkyDome  or maskedSkyDome in a couple of instances. 

I have attached the results of a few tests that I ran. The earlier implementation could not separate the masked and unmasked surfaces successfully in tests 05 and 06. 

Therefore, I have implemented a new method to parse masked and unmasked surfaces. I ran a few tests with this implementation.  It seems to work fine. Following are the results of those tests.

Test - 01 [Off-center shading surface]
![test 01_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462218/aee3db04-af3c-11e6-8830-e38c59ce4942.png)

Test - 02 [A little complex shading surface]
![test 02_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462219/aee64812-af3c-11e6-9244-42521d32bcfb.png)

Test - 03 [Shading surface exactly on the top]
![test 03_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462220/aeed743e-af3c-11e6-84df-d37ee8704c6b.png)

Test - 04 [Shading surface on the top and too close to the test point]
![test 04_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462221/aef02148-af3c-11e6-9d98-fd34c1c43d82.png)

Test - 05 [More than one shading surface]
![test 05_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462222/aef2e270-af3c-11e6-9f92-ba2ca1ee4da9.png)

Test - 06 [Shading surfaces in multiple directions]
![test 06_shading mask ii](https://cloud.githubusercontent.com/assets/20086924/20462223/aef5e54c-af3c-11e6-8704-34093324372f.png)

mostaphaRoudsari commented:
@devngc thanks! Great improvements. Thanks @AntonelloDN for the suggestion.

devngc commented:
@mostaphaRoudsari appreciate it!
